src/hotspot/share/logging/logMessage.hpp

```diff
@@ -62,7 +62,6 @@ template <LogTagType T0, LogTagType T1 = LogTag::__NO_TAG, LogTagType T2 = LogTa
           LogTagType T3 = LogTag::__NO_TAG, LogTagType T4 = LogTag::__NO_TAG, LogTagType GuardTag = LogTag::__NO_TAG>
 class LogMessageImpl : public LogMessageBuffer {
  private:
-  LogImpl<T0, T1, T2, T3, T4, GuardTag> _log;
```

Comment:
I'm not very familiar with the detailed internals here, but don't we need this to capture the Log "state" and ensure it is always the same when used for this LogMessageImpl - otherwise can't `LogTagSetMapping<T0, T1, T2, T3, T4, GuardTag>::tagset()` change what it returns over time?

Reply:
See my comment to Ioi for a more detailed explanation, but in short: No. The template parameters uniquely identify a specific tagset. The tagsets themselves are mutated in place, to accomodate for new outputs and altered decorations, for example.

Reply:
So really `_log` is just a simple alias that makes the other code easier to read IMO. So can't we keep `_log` and change the type?

Reply:
How about `LogTagSet& _tagset`? There are more examples of `_tagset` members in UL, so it's consistent.

Reply:
That seems a bit odd to me. Why can't we use the actual thing we will log to? The description of LogMessage is that it is a multi-part log entry that is accumulated until we flush it by writing to the actual log device.

Reply:
Thanks for that picture! I think my main problem is that `LogTagSet` is actually a much more fundamental abstraction than I was thinking it was from the name. It really embodies a complete entity of logging that I can't readily describe without using words (target, configuration) already used for something else in UL. In that sense `LogImpl` actually did add a level of abstraction that made things a little clearer (and I think `LogImpl` is perhaps what they chose to describe that because they couldn't just use `Log`).
Okay things are a bit clearer now. Thanks.

Reply:
You're welcome! This confusion is exactly why I'm trying to clean things up (I drew the picture for me originally :-)).
Thank you for the review.